### PR TITLE
Add chat templates integration

### PR DESCRIPTION
**Pull request by juliendenize**
This PR aims to add chat templates inside this repo as a valid working implementation in comparison with mistral-common to work with libraries that require jinja templates (llama.cpp, vanilla Transformers, SG Lang ...).

This is in order to give the possibility to our community to use our models beyond the scope of mistral-common when the latter is not integrated in parts of the ecosystem.

We validate these chat templates on a set of dummy requests.

Next step is to add tests for edge cases and errors + integration tests on some datasets.

### CODE REVIEW

File: src/mistral_common/protocol/instruct/messages.py [hunk 1]
```diff
@@ -279,6 +279,8 @@ class ToolMessage(BaseMessage):
     def to_openai(self) -> dict[str, str | list[dict[str, str | dict[str, Any]]]]:
         r"""Converts the message to the OpenAI format."""
         assert self.tool_call_id is not None, "tool_call_id must be provided for tool messages."
+        if self.name is not None:
+            return self.model_dump()
```
Comment (juliendenize) [inline src/mistral_common/protocol/instruct/messages.py]: this is not good should probably revert and add it when need in the tests.

File: tests/data/chat_templates/v15_image.jinja [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+            {{- available_tools }}
+            {{- model_settings }}
+            {%- set ns.available_tools_and_settings_emitted = true %}
+        {%- endif %}
+        {%- if message['content'] is string %}
+            {{- '[INST]' + message['content'] + '[/INST]' }}
+        {%- elif message['content'] | length > 0 %}
+            {{- '[INST]' }}
+            {%- if message['content'] | length == 2 %}
+                {%- set blocks = message['content'] | sort(attribute='type') %}
```
Comment (joa-stdn) [inline tests/data/chat_templates/v15_image.jinja]: is this because of image reordering? written this way is weird no?

File: tests/data/chat_templates/v15_image.jinja [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+{%- set ns_order = namespace(previous_role=none) %}
+{%- for message in loop_messages %}
+    {%- set current_role = message['role'] %}
+    {%- if ns_order.previous_role is not none %}
+        {%- if ns_order.previous_role == 'system' %}
+            {%- if current_role != 'user' and current_role != 'assistant' and current_role != 'system' %}
+                {{- raise_exception('Unexpected role \'' + current_role + '\' after role \'' + ns_order.previous_role + '\'') }}
+            {%- endif %}
+        {%- elif ns_order.previous_role == 'user' %}
+            {%- if current_role != 'assistant' and current_role != 'system' and current_role != 'user' %}
```
Comment (joa-stdn) [inline tests/data/chat_templates/v15_image.jinja]: maybe better to do equality here?

File: tests/data/chat_templates/v15_image.jinja [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+            {%- for msg in ns_agg.current_group %}
+                {%- if msg['content'] is string %}
+                    {%- set ns_c.text_parts = ns_c.text_parts + [msg['content']] %}
+                {%- elif msg['content'] is not none %}
+                    {%- for block in msg['content'] %}
+                        {%- if block['type'] == 'text' %}
+                            {%- set ns_c.text_parts = ns_c.text_parts + [block['text']] %}
+                        {%- else %}
+                            {%- if ns_c.text_parts | length > 0 %}
+                                {%- set ns_c.chunks = ns_c.chunks + [{'type': 'text', 'text': ns_c.text_parts | join('\n\n')}] %}
```
Comment (joa-stdn) [inline tests/data/chat_templates/v15_image.jinja:67]: actually internally we only aggregate across messages now but used indeed to aggregate for chunks in the same message

File: tests/data/chat_templates/v15_image.jinja [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    {%- elif message['role'] == 'system' %}
+        {{- '[SYSTEM_PROMPT]' -}}
+        {%- if message['content'] is string %}
+            {{- message['content'] -}}
+        {%- else %}
+            {%- for block in message['content'] %}
+                {%- if block['type'] == 'text' %}
+                    {{- block['text'] }}
+                {%- else %}
+                    {{- raise_exception('Only text chunks are supported in system message contents.') }}
```
Comment (joa-stdn) [inline tests/data/chat_templates/v15_image.jinja]: do you think we could have a macro for some of those things? like render_content?

File: tests/data/chat_templates/v15_image_think.jinja [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+        {%- if message['content'] is string and message['content'] != '' %}
+            {{- message['content'] }}
+        {%- elif message['content'] | length > 0 %}
+            {%- for block in message['content'] %}
+                {%- if block['type'] == 'text' %}
+                    {{- block['text'] }}
+                {%- elif block['type'] == 'thinking' %}
+                    {{- '[THINK]' + block['thinking'] }}
+                    {%- if block.get('closed', true) %}{{- '[/THINK]' }}{%- endif %}
```
Comment (joa-stdn) [inline tests/data/chat_templates/v15_image_think.jinja]: tbh it should really always be true IMO, are we sure it's going to be ok?

Comment (juliendenize) [reply tests/data/chat_templates/v15_image_think.jinja]: it's an option from mistral-common so i'd like to replicate, think it should be fine anyway as i think very few people would pass this flag.

File: .github/workflows/lint_build_test.yaml [hunk 7]
```diff
@@ -95,3 +95,32 @@ jobs:
...
+      matrix:
+        python-version: [
+          "3.10",
+          "3.11",
+          "3.12",
+          "3.13",
+          "3.14",
+        ]
+    steps:
+      - name: Checkout
```
Comment (joa-stdn) [inline .github/workflows/lint_build_test.yaml:112]: are we ok with github actions?

File: src/mistral_common/integrations/chat_templates/chat_templates.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+from mistral_common.tokens.tokenizers.base import TokenizerVersion
+
+
+def generate_chat_template(
+    spm: bool,
+    tokenizer_version: TokenizerVersion,
+    image_support: bool,
+    audio_support: bool,
+    thinking_support: bool,
+    default_system_prompt: str | None = None,
```
Comment (joa-stdn) [inline src/mistral_common/integrations/chat_templates/chat_templates.py]: let's remove defaults?

File: src/mistral_common/integrations/chat_templates/chat_templates.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        image_support=image_support,
+        audio_support=audio_support,
+        thinking_support=thinking_support,
+        plain_thinking_support=plain_thinking_support,
+        use_token_variables=use_token_variables,
+    )
+    template = _build_chat_template(config)
+
+    if default_system_prompt is not None:
+        escaped_prompt = default_system_prompt.replace("\\", "\\\\").replace("'", "\\'")
```
Comment (joa-stdn) [inline src/mistral_common/integrations/chat_templates/chat_templates.py:56]: we're not afraid of the thinking chunk here? Worth a comment at least maybe?

File: src/mistral_common/integrations/chat_templates/template_generator.py [hunk 10]
```diff
@@ -0,0 +1,1674 @@
...
+        True
+    """
+
+    version: TokenizerVersion
+    spm: bool = False
+    image_support: bool = False
+    audio_support: bool = False
+    thinking_support: bool = False
+    plain_thinking_support: bool = False
+    use_token_variables: bool = True
```
Comment (joa-stdn) [inline src/mistral_common/integrations/chat_templates/template_generator.py]: this name is very weird, I wonder if we can maybe find a better name no?

File: src/mistral_common/integrations/chat_templates/template_generator.py [hunk 11]
```diff
@@ -0,0 +1,1674 @@
...
+            raise ValueError("Plain thinking support is only available for tokenizer version v11")
+        if self.audio_support and self.plain_thinking_support:
+            raise ValueError("Audio and plain thinking support are mutually exclusive")
+
+    @property
+    def bos_expr(self) -> str:
+        r"""Jinja expression for the BOS token."""
+        if self.use_token_variables:
+            return "bos_token"
+        return "'" + _BOS + "'"
```
Comment (joa-stdn) [inline src/mistral_common/integrations/chat_templates/template_generator.py]: is this better than f strings?

File: src/mistral_common/integrations/chat_templates/template_generator.py [hunk 12]
```diff
@@ -0,0 +1,1674 @@
...
+                "        {{- raise_exception('Only user, assistant, system and tool roles are supported, got ' + message['role'] + '.') }}"  # noqa: E501
+            )
+        else:
+            lines.append(
+                "        {{- raise_exception('Only user, assistant and system roles are supported, got ' + message['role'] + '.') }}"  # noqa: E501
+            )
+    else:
+        # pre-v7: system messages are already filtered out
+        if config.has_tools:
+            if config.version == TokenizerVersion.v2 or (config.spm and config.version == TokenizerVersion.v3):
```
Comment (joa-stdn) [inline src/mistral_common/integrations/chat_templates/template_generator.py]: to me it's quite weird to have both explicit versions and the simplifying properties, we should go one way or the other no?

File: tests/integrations/chat_templates/transformers/test_core_parity.py [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+            spm=spm, tokenizer_version=version, image_support=image, audio_support=audio, thinking_support=think
+        )
+        if version <= TokenizerVersion.v2:
+            for conv in conversations:
+                for message in conv.messages:
+                    if isinstance(message, (UserMessage, AssistantMessage)) and isinstance(message.content, list):
+                        assert len(message.content) == 1 and isinstance(message.content[0], TextChunk), (
+                            "Only text content is supported for v1 and v2"
+                        )
+                        message.content = str(message.content[0].text)
```
Comment (joa-stdn) [inline tests/integrations/chat_templates/transformers/test_core_parity.py:74]: is it better to do this or to have them already in str?

File: tests/integrations/chat_templates/transformers/test_core_parity.py [hunk 14]
```diff
@@ -0,0 +1,310 @@
...
+        for conversation in conversations:
+            if version == TokenizerVersion.v2:
+                for message in conversation.messages:
+                    if isinstance(message, ToolMessage):
+                        message.name = "tool"
```
Comment (joa-stdn) [inline tests/integrations/chat_templates/transformers/test_core_parity.py]: I think we can just keep names all the time?

File: tests/integrations/chat_templates/transformers/test_core_parity.py [hunk 15]
```diff
@@ -0,0 +1,310 @@
...
+        # Consecutive user messages should be aggregated (not raise an error)
+        valid_consecutive_users = {
+            "messages": [
+                {"role": "user", "content": "Hello"},
+                {"role": "user", "content": "World"},
+                {"role": "assistant", "content": "Hi"},
+            ]
+        }
+        # This should not raise
+        encode_transformers(chat_template, valid_consecutive_users)
```
Comment (joa-stdn) [inline tests/integrations/chat_templates/transformers/test_core_parity.py]: why don't we check matching here?

File: tests/integrations/chat_templates/transformers/test_core_parity.py [hunk 16]
```diff
@@ -0,0 +1,310 @@
...
+    def test_invalid_chunks(
+        self,
+        spm: bool,
+        version: TokenizerVersion,
+        image: bool,
+        audio: bool,
+        think: bool,
+    ) -> None:
+        invalid_sp_think = {
+            "messages": [
```
Comment (joa-stdn) [inline tests/integrations/chat_templates/transformers/test_core_parity.py]: maybe let's define this in fixtures / outside of tests and have the test itself shorter?

File: tests/integrations/chat_templates/transformers/test_core_parity.py [hunk 17]
```diff
@@ -0,0 +1,310 @@
...
+        spm: bool,
+        version: TokenizerVersion,
+        mode: ValidationMode,
+        image: bool,
+        audio: bool,
+        think: bool,
+    ) -> None:
+        conversations = _get_conversations(version, mode, image, audio, think)
+
+        mistral_tokenizer = _get_mistral_tokenizer(
```
Comment (joa-stdn) [inline tests/integrations/chat_templates/transformers/test_core_parity.py]: curious why we don't do the MistralTokenizer.from_file directly with a file in tmp_path etc?

File: tests/integrations/chat_templates/conftest.py [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+from unittest.mock import patch
+
+import pytest
+
+from mistral_common.tokens.tokenizers.base import TokenizerVersion
+from tests.integrations.chat_templates.helpers import _IMAGE
+
+# All configurations for output comparison tests (including SPM).
+# Tuples: (version, spm, image, audio, think, plain_think)
+ALL_CONFIGS: list[tuple[TokenizerVersion, bool, bool, bool, bool, bool]] = [
```
Comment (joa-stdn) [inline tests/integrations/chat_templates/conftest.py]: should we dataclass for this?

File: tests/integrations/chat_templates/helpers.py [hunk 19]
```diff
@@ -0,0 +1,340 @@
...
+        render_kwargs["reasoning_effort"] = reasoning_effort
+
+    return jinja_template.render(**render_kwargs)
+
+
+def encode_mistral_common(mistral_tokenizer: MistralTokenizer, chat_request: ChatCompletionRequest, spm: bool) -> str:
+    r"""Encode a chat request using mistral-common tokenizer."""
+    mistral_encoded = str(mistral_tokenizer.encode_chat_completion(chat_request).text)
+    # Remove image tokens except one per image
+    mistral_encoded = mistral_encoded.replace("[IMG]", "").replace("[IMG_BREAK]", "").replace("[IMG_END]", "[IMG]")
```
Comment (joa-stdn) [inline tests/integrations/chat_templates/helpers.py:127]: Actually doesn't it make more sense to concatenate all [IMG] into [IMG] and drop the two others? Maybe we don't care, this is nit

File: tests/integrations/chat_templates/helpers.py [hunk 20]
```diff
@@ -0,0 +1,340 @@
...
+    # Only add reasoning_effort for v15+ templates that support it
+    if reasoning_effort is not None or "reasoning_effort" in template:
+        render_kwargs["reasoning_effort"] = reasoning_effort
+
+    return jinja_template.render(**render_kwargs)
+
+
+def encode_mistral_common(mistral_tokenizer: MistralTokenizer, chat_request: ChatCompletionRequest, spm: bool) -> str:
+    r"""Encode a chat request using mistral-common tokenizer."""
+    mistral_encoded = str(mistral_tokenizer.encode_chat_completion(chat_request).text)
```
Comment (joa-stdn) [inline tests/integrations/chat_templates/helpers.py:123]: there's no way to compare token ids? I thought this was the purpose...

File: tests/integrations/chat_templates/helpers.py [hunk 21]
```diff
@@ -0,0 +1,340 @@
...
+
+    render_kwargs: dict[str, Any] = {
+        "messages": messages,
+        "tools": tools,
+        "bos_token": "<s>",
+        "eos_token": "</s>",
+    }
+
+    # Only add reasoning_effort for v15+ templates that support it
+    if reasoning_effort is not None or "reasoning_effort" in template:
```
Comment (joa-stdn) [inline tests/integrations/chat_templates/helpers.py]: seems weird to me, I would have said that we need to restructure the tests so that this line can just be
```
if reasoning_effort is not None
```
no?

File: tests/integrations/chat_templates/helpers.py [hunk 22]
```diff
@@ -0,0 +1,340 @@
...
+    mistral_encoded = mistral_encoded.replace("[IMG]", "").replace("[IMG_BREAK]", "").replace("[IMG_END]", "[IMG]")
+    # Remove audio tokens except one per audio
+    mistral_encoded = mistral_encoded.replace("[AUDIO]", "").replace("[BEGIN_AUDIO]", "[AUDIO]")
+    if spm:
+        mistral_encoded = mistral_encoded.replace(SPM_WHITESPACE, " ").replace("<0x0A>", "\n")
+    return mistral_encoded
+
+
+def encode_transformers(
+    chat_template: str, chat_request: ChatCompletionRequest | dict[str, Any], keep_name_for_tools: bool = False
```
Comment (joa-stdn) [inline tests/integrations/chat_templates/helpers.py]: why do we allow both pydantic object and ChatCompletionRequest

File: tests/integrations/chat_templates/helpers.py [hunk 23]
```diff
@@ -0,0 +1,340 @@
...
+        openai_request = chat_request
+    for tool in openai_request.get("tools", []):
+        tool["function"].pop("strict", False)
+
+    reasoning_effort = openai_request.get("reasoning_effort")
+    template_kwargs: dict[str, Any] = {}
+    if reasoning_effort is not None:
+        template_kwargs["reasoning_effort"] = reasoning_effort
+
+    encoded = render_jinja_template(
```
Comment (joa-stdn) [inline tests/integrations/chat_templates/helpers.py]: maybe add a comment on why we have render_template as well?

File: tests/integrations/chat_templates/helpers.py [hunk 24]
```diff
@@ -0,0 +1,340 @@
...
+            return InstructTokenizerV11
+        case TokenizerVersion.v13:
+            return InstructTokenizerV13
+        case TokenizerVersion.v15:
+            return InstructTokenizerV15
+        case _:
+            raise ValueError(f"Unknown tokenizer version: {tokenizer_version}")
+
+
+def _get_mistral_tekkenizer(
```
Comment (joa-stdn) [inline tests/integrations/chat_templates/helpers.py]: why don't we use MistralTokenizer.from_file?

File: tests/integrations/chat_templates/helpers.py [hunk 25]
```diff
@@ -0,0 +1,340 @@
...
+    tokenizer_version: TokenizerVersion, validation_mode: ValidationMode, image: bool, audio: bool, think: bool
+) -> MistralTokenizer:
+    r"""Build a `MistralTokenizer` with Tekken backend."""
+    special_tokens = get_special_tokens(tokenizer_version=tokenizer_version, add_audio=audio, add_think=think)
+    with open(MistralTokenizer._data_path() / "tekken_240911.json", "r", encoding="utf-8") as f:
+        json_tekkenizer = json.load(f)
+    vocab = json_tekkenizer["vocab"]
+    vocab_size = json_tekkenizer["config"]["default_vocab_size"]
+    pattern = json_tekkenizer["config"]["pattern"]
+    model_settings_builder = (
```
Comment (joa-stdn) [inline tests/integrations/chat_templates/helpers.py]: this is too hacky no?

File: tests/data/chat_templates/v15_image.jinja (hunk 4 above)
Comment (juliendenize) [reply tests/data/chat_templates/v15_image.jinja:67]: should be something that i change in the end but should be first deprecated

File: .github/workflows/lint_build_test.yaml (hunk 7 above)
Comment (juliendenize) [reply .github/workflows/lint_build_test.yaml:112]: i should move to bk probably eventually. will adress in followup

File: src/mistral_common/integrations/chat_templates/chat_templates.py (hunk 9 above)
Comment (juliendenize) [reply src/mistral_common/integrations/chat_templates/chat_templates.py:56]: wdym ? think chunks doesn't need escape right ?
Or do you mean that users should keep in mind to add think tokens if needed ?

File: src/mistral_common/integrations/chat_templates/template_generator.py (hunk 10 above)
Comment (juliendenize) [reply src/mistral_common/integrations/chat_templates/template_generator.py]: changed the name lmk if it looks better :)

File: src/mistral_common/integrations/chat_templates/template_generator.py (hunk 11 above)
Comment (juliendenize) [reply src/mistral_common/integrations/chat_templates/template_generator.py]: nop nice catch

File: src/mistral_common/integrations/chat_templates/template_generator.py (hunk 12 above)
Comment (juliendenize) [reply src/mistral_common/integrations/chat_templates/template_generator.py]: true, added more properties

File: tests/integrations/chat_templates/transformers/test_core_parity.py (hunk 14 above)
Comment (juliendenize) [reply tests/integrations/chat_templates/transformers/test_core_parity.py]: true

File: tests/integrations/chat_templates/transformers/test_core_parity.py (hunk 15 above)
Comment (juliendenize) [reply tests/integrations/chat_templates/transformers/test_core_parity.py]: these tests are redundant because already tested valid ones in the conversations, going to remove.

File: tests/integrations/chat_templates/transformers/test_core_parity.py (hunk 17 above)
Comment (juliendenize) [reply tests/integrations/chat_templates/transformers/test_core_parity.py]: we don't have all the versions in the fixtures data. I could download some official but decided instead to build them on runtime. Is it a blocker to you ?

File: tests/integrations/chat_templates/helpers.py (hunk 20 above)
Comment (juliendenize) [reply tests/integrations/chat_templates/helpers.py:123]: it is possible but requires to use apply_chat_template from HF. I could add this indeed.

File: tests/integrations/chat_templates/helpers.py (hunk 22 above)
Comment (juliendenize) [reply tests/integrations/chat_templates/helpers.py]: the other one was to test invalid or dict request from openai, created another function

File: tests/integrations/chat_templates/helpers.py (hunk 25 above)
Comment (juliendenize) [reply tests/integrations/chat_templates/helpers.py]: why do you think that ?

File: tests/integrations/chat_templates/transformers/test_core_parity.py (hunk 17 above)
Comment (juliendenize) [reply tests/integrations/chat_templates/transformers/test_core_parity.py]: in the end i changed it to be able to use MistralCommonBackend :)

File: tests/integrations/chat_templates/helpers.py (hunk 20 above)
Comment (juliendenize) [reply tests/integrations/chat_templates/helpers.py:123]: added

File: tests/integrations/chat_templates/transformers/test_core_parity.py (hunk 13 above)
Comment (juliendenize) [reply tests/integrations/chat_templates/transformers/test_core_parity.py:74]: better like this as i also want to test chunks

File: tests/integrations/chat_templates/helpers.py (hunk 19 above)
Comment (juliendenize) [reply tests/integrations/chat_templates/helpers.py:127]: harder to setup if you have multiple images that are one after the other

File: tests/integrations/chat_templates/helpers.py (hunk 23 above)
Comment (juliendenize) [reply tests/integrations/chat_templates/helpers.py]: there is a docstring for the function that should explain it

File: src/mistral_common/integrations/chat_templates/chat_templates.py (hunk 9 above)
Comment (joa-stdn) [reply src/mistral_common/integrations/chat_templates/chat_templates.py:56]: I meant system prompt themselves can contain chunks? But here it's str, should be ok

File: src/mistral_common/integrations/chat_templates/template_generator.py [hunk 26]
```diff
@@ -0,0 +1,1698 @@
...
+    ]
+
+
+def _generate_message_aggregation(config: TemplateConfig) -> str:
+    r"""Generate message aggregation pre-processing block.
+
+    Aggregates consecutive messages with the same role before the alternation
+    check to match the behavior of the mistral-common normalizer:
+
+    - **User messages**: consecutive messages merged into one, text content joined
```
Comment (joa-stdn) [inline src/mistral_common/integrations/chat_templates/template_generator.py:685]: we need to nuke this actually, any chance we can do that in follow up?

File: src/mistral_common/integrations/chat_templates/template_generator.py [hunk 27]
```diff
@@ -0,0 +1,1698 @@
...
+      with `"\\n\\n"`, non-text chunks (image, audio) preserved in order.
+    - **Assistant messages**: consecutive messages merged into one, text content joined
+      with `"\\n\\n"`, `tool_calls` lists concatenated.
```
Comment (joa-stdn) [inline src/mistral_common/integrations/chat_templates/template_generator.py:688]: all of this is no longer up to date in new models I think

File: tests/integrations/chat_templates/helpers.py [hunk 28]
```diff
@@ -0,0 +1,695 @@
...
+    audio_encoder = _get_audio_encoder() if audio else None
+    image_encoder = _get_image_encoder(tokenizer) if image else None
+
+    instruct_tokenizer: InstructTokenizer = _get_instruct_tokenizer_class(tokenizer_version)(
+        tokenizer=tokenizer,
+        image_encoder=image_encoder,
+        audio_encoder=audio_encoder,
+    )
+    model_settings_builder = tokenizer.model_settings_builder if isinstance(tokenizer, Tekkenizer) else None
+    return MistralTokenizer(
```
Comment (joa-stdn) [inline tests/integrations/chat_templates/helpers.py]: I think MistralTokenizer.from_file is more natural, we have that in test_tokenizer_v14 etc no?

File: tests/integrations/chat_templates/helpers.py [hunk 29]
```diff
@@ -0,0 +1,695 @@
...
+
+    audio_encoder = _get_audio_encoder() if audio else None
+    image_encoder = _get_image_encoder(tokenizer) if image else None
+
+    instruct_tokenizer: InstructTokenizer = _get_instruct_tokenizer_class(tokenizer_version)(
+        tokenizer=tokenizer,
+        image_encoder=image_encoder,
+        audio_encoder=audio_encoder,
+    )
+    model_settings_builder = tokenizer.model_settings_builder if isinstance(tokenizer, Tekkenizer) else None
```
Comment (joa-stdn) [inline tests/integrations/chat_templates/helpers.py]: this would allow not doing this

File: tests/integrations/chat_templates/helpers.py [hunk 30]
```diff
@@ -0,0 +1,695 @@
...
+
+    for attr_name in vars(instruct_tokenizer):
+        val = getattr(instruct_tokenizer, attr_name)
+        if isinstance(val, int) and 0 <= val < num_special:
+            active_ids.add(val)
+
+    return active_ids
+
+
+def _build_hf_tokenizer(
```
Comment (joa-stdn) [inline tests/integrations/chat_templates/helpers.py]: can do we a file for hf_utils maybe? This one is quite big?

File: tests/integrations/chat_templates/helpers.py [hunk 31]
```diff
@@ -0,0 +1,695 @@
...
+    """
+    tokenizer = instruct_tokenizer.tokenizer
+    # BOS and EOS are always active
+    active_ids: set[int] = {tokenizer.bos_id, tokenizer.eos_id}
+
+    num_special = getattr(tokenizer, "num_special_tokens", 0)
+    if num_special == 0:
+        return active_ids
+
+    for attr_name in vars(instruct_tokenizer):
```
Comment (joa-stdn) [inline tests/integrations/chat_templates/helpers.py]: not super clean no? why don't we just add all special tokens to the HF model? this is what we do anyways in Tekkenizer?

File: tests/integrations/chat_templates/helpers.py [hunk 32]
```diff
@@ -0,0 +1,695 @@
...
+    r"""Encode a chat request using the transformers render_jinja_template.
+
+    Converts a ChatCompletionRequest to OpenAI format and renders through
+    the HuggingFace transformers rendering pipeline.
+    """
+    assert _HAS_TRANSFORMERS, "transformers is required"
+    openai_request = chat_request.to_openai()
+    if keep_name_for_tools:
+        for openai_message, chat_message in zip(openai_request["messages"], chat_request.messages):
+            if chat_message.role == "tool":
```
Comment (joa-stdn) [inline tests/integrations/chat_templates/helpers.py]: this is duplicated can we factorize?

File: tests/integrations/chat_templates/helpers.py [hunk 33]
```diff
@@ -0,0 +1,384 @@
...
+        tokenizer_version,
+    )
+
+    image_encoder = _get_image_encoder(tokenizer) if image else None
+
+    instruct_tokenizer: InstructTokenizer = _get_instruct_tokenizer_class(tokenizer_version)(
+        tokenizer=tokenizer,
+        image_encoder=image_encoder,
+    )
+    return MistralTokenizer(
```
Comment (joa-stdn) [inline tests/integrations/chat_templates/helpers.py]: why not from file here?

**Review by joa-stdn — APPROVED** [review]

File: tests/integrations/chat_templates/helpers.py (hunk 25 above)
Comment (juliendenize) [reply tests/integrations/chat_templates/helpers.py]: Refactored — now uses `MistralTokenizer.from_file` via `_build_tekken_json` + `_get_mistral_tekkenizer`. No more manual tokenizer construction.
